### ISSUE TITLE
path/filepath: filepath.Walk does not work on mounted volumes in Windows containers

### ISSUE THREAD

**Issue by buddhike** [golang/go#26033]
### What version of Go are you using (`go version`)?
1.10

### Does this issue reproduce with the latest release?
Not sure

### What operating system and processor architecture are you using (`go env`)?
`windows 386`

### What did you do?
- Cross compile the following program from osx `GOOS=windows GOARCH=386 go build -o walk.exe main.go` (cross compilation does not really matter, the behaviour outlined below is also consistent even if the binary is produced in a windows)

- Created a docker container from `microsoft/windowsservercore`
```
docker run --rm -it -v c:\files:c:\files microsoft/windowsservercore
```

- Ran the program on the mounted volume
```
walk.exe -path c:\files
```

```
package main

import (
	"flag"
	"fmt"
	"os"
	"path/filepath"
)

func main() {
	path := flag.String("path", "", "path to walk")
	flag.Parse()
	fmt.Printf("walking path %v\n", *path)

	err := filepath.Walk(*path, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			panic(err)
		}
		fmt.Printf("visiting %v\n", path)
		return nil
	})

	if err != nil {
		panic(err)
	}
}
```

### What did you expect to see?
List of files and directories in `c:\files`.

### What did you see instead?
Just the root directory

**Comment by slrz** [comment]
You run the walk program outside of the started container, just on the same directory that is also mounted into the container?

The directory search works ok when using other (non-Go) programs? Can you check what [this C program](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365200(v=vs.85).aspx) would output?

**Comment by buddyspike** [comment]
Sorry about the lag in my response. Program is run from the default CWD inside the container. However it's in a mounted directory. To be precise:

Step 1 - Launch the container. `c:\files` has the items we want to traverse. `c:\bin` has `walk.exe`.
```
docker run --rm -it -v c:\files:c:\files -v c:\bin:c:\bin microsoft/windowsservercore
```

Step 2 - From `cwd`, run `walk.exe`
```
c:\bin\walk.exe -path c:\files
```

Using other programs such as `dir` works just fine on mounted directories.
I will update this thread the with the result of the above program over the next couple of days. 

Hopefully this clarifies the situation a bit further.

**Comment by igorrendulic** [comment]
same issue with mounted docker volume on linux. If i run it from VSCode/command line it works as expected (on the same folder which is not mounted). As long as it's mounted through docker volume, it doesn't work. Logged into docker container to check if folder with files exists inside of it and all is there. Might be permission issues of some sorts?

**Comment by TBBle** [comment]
I've confirmed the same behaviour with go1.15.5 for junction points outside a container. My use-case is for volume mounts that happen to _be_ a WCOW container image.

I mounted a WCOW container image using `wclayer`, and used the source code in the initial report to test. Walking inside the mount point is fine, but walking the mount point directly or walking a directory above the mount point, the walk stops at the mount.

I suspect the core problem is that `filepath.Walk` does not follow symbolic links, and to Go, a volume mount point appears to be a symlink back to itself, see https://github.com/containerd/containerd/pull/4419/commits/ac88af7a058e0e724e7e972293284f0768319d07#r485615803 and `isSymlink` in https://golang.org/src/os/types_windows.go.

Unlike a real Symlink, it's not necessarily feasible to call `filepath.Walk` on the target of a mount point, as it may be a volume reference.

**Comment by TBBle** [comment]
Thinking about this, one rationale for treating an NTFS Mount Point like a Symlink is that, unlike `mount --rbind` in Linux, subsequently-created Volume Mounts are visible through that reparse point, so it's possible to introduce an infinite recursion.

I assume that avoiding infinite recursion is why `filepath.Walk` doesn't follow symlinks in the first place.

**Comment by bcmills** [comment]
After https://go.dev/cl/463177, the fix here is probably to add a trailing separator to the path passed to `Walk`.

**Comment by TBBle** [comment]
For the record, the above fix was in Go 1.21.
